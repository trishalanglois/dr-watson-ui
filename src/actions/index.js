export const createUser = user => ({
  type: 'CREATE_USER',
  user
});

export const removeUser = () => ({
  type: 'REMOVE_USER'
});

export const hasErrored = errorMsg => ({
  type: 'HAS_ERRORED',
  errorMsg
})

export const addMessage = message => ({
  type: 'ADD_MESSAGE',
  message
})

export const clearMessages = () => ({
  type: 'CLEAR_MESSAGES',
  messages: []
})
