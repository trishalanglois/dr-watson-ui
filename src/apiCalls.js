export const startConversation = async feeling => {
  const url = 'https://drwatson-api.herokuapp.com/api/v1/start_session'
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ feeling })
  };

  const response = await fetch(url, options);
  if (!response.ok) {
    throw Error('Dr Watson is currently down.  Please try again later.')
  }

  const data = await response.json();
  console.log(data)
  return data;
}

export const postMessage = async newMessage => {
  const url = 'https://drwatson-api.herokuapp.com/api/message'
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ newMessage })
  };

  const response = await fetch(url, options);
  if (!response.ok) {
    throw Error('Dr Watson is currently down.  Please try again later.')
  }

  const data = await response.json();
  console.log(data);
  return data;
}

export const endConversation = async () => {
  const url = 'https://drwatson-api.herokuapp.com/api/v1/end_session';

  const response = await fetch(url);
  if (!response.ok) {
    throw Error('There was a problem ending the session.  Please close the application.')
  }
}
