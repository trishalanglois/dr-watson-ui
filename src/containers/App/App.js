import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Header from '../Header/Header';
import WelcomeModal from '../WelcomeModal/WelcomeModal';
import ChatBox from '../ChatBox/ChatBox';
import { removeUser, hasErrored, clearMessages } from '../../actions';
import { endConversation } from '../../apiCalls';
import './App.css';

export class App extends Component {
  constructor(props) {
    super(props);
  }

  signOut = async () => {
    try {
      await endConversation();
      this.props.removeUser();
      this.props.clearMessages();
    } catch({ message }) {
      this.props.hasErrored(message);
    }
  }

  render() {
    const { user } = this.props;
    return (
      <div className="App">
        <Header signOut={this.signOut} />
        {!user && <WelcomeModal />}
        {user && <ChatBox />}
      </div>
    );
  }
}

export const mapStateToProps = ({ user }) => ({
  user
});

export const mapDispatchToProps = dispatch =>  bindActionCreators({
  removeUser,
  hasErrored,
  clearMessages
 }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(App);
