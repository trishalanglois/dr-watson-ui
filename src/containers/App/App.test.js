import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { removeUser, hasErrored } from '../../actions';
import { endConversation } from '../../apiCalls';

jest.mock('../../apiCalls');

describe('App component', () => {
  const mockRemoveUser = jest.fn();
  const mockHasErrored = jest.fn();
  let wrapper;

  beforeEach(() => {
    const mockUser = {
      id: 1568665187737,
      firstName: "Travis",
      lastName: "Rollins",
      feeling: "tired"
    };

    wrapper = shallow(<App
        user={mockUser}
        removeUser={mockRemoveUser}
        hasErrored={mockHasErrored}
    />);
  });

  it('should match the snapshot with a ChatBox if the user has signed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should match the snapshot with a WelcomeModal if the user hasn't signed in yet", () => {
    const wrapper = shallow(<App
      user={null}
      removeUser={mockRemoveUser}
      hasErrored={mockHasErrored}
    />);

    expect(wrapper).toMatchSnapshot();
  });


  it('should call endConversation, and removeUser if someone signs out', async () => {
    await wrapper.instance().signOut();

    expect(endConversation).toHaveBeenCalled();
    expect(mockRemoveUser).toHaveBeenCalled();
  });

  it('should call hasErrored if endCoversation does not resolve when a user signs out', async () => {
    endConversation.mockImplementation(() => {
      return Promise.reject(Error('fetch failed.'));
    });

    await wrapper.instance().signOut();

    expect(mockHasErrored).toHaveBeenCalledWith('fetch failed.');
  });
});

describe('mapStateToProps', () => {
  it('should return an object with the user information', () => {
    const mockUser = {
      id: 1568665187737,
      firstName: "Travis",
      lastName: "Rollins",
      feeling: "tired"
    };

    const mockState = {
      user: mockUser,
      messages: [{
        message: 'Hi there, my name is Dr. Watson. I understand that you have been feeling happy. That is super exciting to hear!',
        isUser: false,
      }],
      errorMsg: ''
    };
    const expected = {
      user: mockUser
    }

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  it('calls dispatch with a removeUser action when removeUser is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = removeUser();

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeUser();

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls dispatch with a hasErrored action when hasErrored is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = hasErrored('fetch failed');

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.hasErrored('fetch failed');

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  if('calls dispatch with a clearMessages action when clearMessages is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = clearMessages();

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.clearMessages();

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
  
});
