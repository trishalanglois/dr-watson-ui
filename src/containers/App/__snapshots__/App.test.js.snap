// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot with a ChatBox if the user has signed in 1`] = `
<div
  className="App"
>
  <Connect(Header)
    signOut={[Function]}
  />
  <Connect(ChatBox) />
</div>
`;

exports[`App component should match the snapshot with a WelcomeModal if the user hasn't signed in yet 1`] = `
<div
  className="App"
>
  <Connect(Header)
    signOut={[Function]}
  />
  <Connect(WelcomeModal) />
</div>
`;
